contracts/delegation/dao-vote-delegation/src/hooks.rs

```diff
@@ -0,0 +1,342 @@
+                    // block and update the total that will be reflected in
+                    // historical queries starting from the next block.
+                    env.block.height + 1,
+                    delta,
+                )?;
+
+                // perform the inverse of the change on the expiration block. do
+                // this decrement after the increment to avoid underflow.
+                if let Some(expiration) = expiration {
+                    DELEGATED_VP.decrement(deps.storage, delegate.clone(), expiration, delta)?;
+                }
+            }
+        }
+    }
+
+    Ok(Response::new()
+        .add_attribute("action", "voting_power_change_hook")
+        .add_attribute("member_type", "delegator"))
+}
+
+// if first vote by a delegator, update the unvoted delegated VP for their
+// delegates, if any, by subtracting this member's delegated VP. if not first
+// vote, this has already been done.
+pub fn execute_vote_hook(
+    deps: DepsMut,
+    proposal_module: Addr,
+    vote_hook: VoteHookMsg,
+) -> Result<Response, ContractError> {
+    // ensure proposal module is registered
+    if !PROPOSAL_HOOK_CALLERS.has(deps.storage, proposal_module.clone()) {
+        return Err(ContractError::UnauthorizedHookCaller {});
+    }
+
+    match vote_hook {
+        VoteHookMsg::NewVote {
+            proposal_id,
+            voter,
+            power,
+            height,
+            is_first_vote,
+            ..
+        } => {
+            // if not first vote, this has already been done.
+            if !is_first_vote {
+                return Ok(Response::new()
+                    .add_attribute("action", "vote_hook")
+                    .add_attribute("is_first_vote", "false"));
+            }
+
+            // update voting power for all delegates, if any. if this voter is a
+            // delegate themself, there will simply be no delegations.
+            let delegator = deps.api.addr_validate(&voter)?;
+            let delegations = DELEGATIONS.load_all(deps.storage, &delegator, height)?;
+            for LoadedItem {
+                item: Delegation { delegate, percent },
+                ..
+            } in delegations
+            {
+                let udvp = get_udvp(
+                    deps.as_ref(),
+                    &delegate,
+                    &proposal_module,
+                    proposal_id,
+                    height,
+                )?;
+
+                let delegated_vp = calculate_delegated_vp(power, percent);
+
+                // remove the delegator's delegated VP from the delegate's
+                // unvoted delegated VP for this proposal since this
+                // delegator just voted.
+                let new_udvp = udvp.checked_sub(delegated_vp)?;
+
+                UNVOTED_DELEGATED_VP.save(
+                    deps.storage,
+                    (&delegate, &proposal_module, proposal_id),
+                    &new_udvp,
+                )?;
```

Comment:
This seems to be the only thing we need to ensure that each unit of vp is only counted once, right?

Reply:
yep, UDVP is the source of truth for how much voting power a delegate should have tallied in a proposal. at the end of every TX, a proposal ballot for a delegate should be their personal VP + their UDVP value (with the cap applied if any)